### PR TITLE
FFI for Java's CompletionStage

### PR DESCRIPTION
**Pull request by calvinlfer**
Took a stab at completing https://github.com/getkyo/kyo/issues/323

### CODE REVIEW

File: kyo-core/jvm/src/test/scala/kyoTest/completionStageTest.scala [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package kyoTest
+
+import java.util.concurrent.*
+import kyo.*
+import kyo.Fibers
+
+class completionStageTest extends KyoTest:
+    "completionStage to Fiber" in runJVM {
```
Comment (fwbrasil) [inline kyo-core/jvm/src/test/scala/kyoTest/completionStageTest.scala:8]: Could you add a test with a failure and another with locals? For locals, maybe something along these lines:

```scala
val local = Locals.init[Int](0)
Locals.let(42) {
   Fibers.fromCompletionStage(...).map { _ =>
       assert(local.get == 42)
   }
}
```

Comment (calvinlfer) [reply kyo-core/jvm/src/test/scala/kyoTest/completionStageTest.scala:8]: yessir

File: kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package kyo
+
+import java.util.concurrent.CompletionStage
+import kyo.scheduler.IOPromise
+import kyo.scheduler.IOTask
+
+trait fibersPlatformSpecific:
```
[the remark this answers is not in the document]
Comment (fwbrasil) [reply kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala:7]: I think if the extension exposes the methods correctly and doesn't have type inference issues, which should be the case, it'd be a better solution than the trait extension. 👍🏽

[the remark this answers is not in the document]
Comment (calvinlfer) [reply kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala:7]: I'm running into an issue here:
```scala
extension (fibers: Fibers.type)
    def fromCompletionStage[T: Flat](cs: CompletionStage[T]): T < Fibers =
        Fibers.get(fromCompletionStageFiber(cs))

    def fromCompletionStageFiber[T: Flat](cs: CompletionStage[T]): Fiber[T] < IOs =
        Locals.save { st =>
            IOs {
                val p = new IOPromise[T]()
                cs.whenComplete { (success, error) =>
                    val io = IOs {
                        if error == null then p.complete(success)
                        else p.complete(IOs.fail(error))
                    }
                    IOTask(io, st)
                    ()
                }
                Promise(p)
            }
        }
end extension
```

```json
[{
	"resource": "/Users/cal/IdeaProjects/kyo/kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala",
	"owner": "_generated_diagnostic_collection_name_#3",
	"severity": 8,
	"message": "Could not read definition type Fibers in /Users/cal/IdeaProjects/kyo/.bloop/kyo-core/bloop-internal-classes/classes-Metals-OASR1p6JRI6PENVr-qmJTw==-aUogg80sQ2iR066wjsQS8w==/kyo/fibers$package.tasty. Caused by the following exception:\njava.lang.AssertionError: assertion failed: module class Fibers$ has non-class parent: TypeRef(TermRef(ThisType(TypeRef(NoPrefix,module class <root>)),object kyo),fibersPlatformSpecific)\n\nRun with -Ydebug-unpickling to see full stack trace.",
	"source": "bloop",
	"startLineNumber": 8,
	"startColumn": 67,
	"endLineNumber": 8,
	"endColumn": 67
}]
```

[the remark this answers is not in the document]
Comment (fwbrasil) [reply kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala:7]: I see, I think the trait approach is fine then!

[the remark this answers is not in the document]
Comment (calvinlfer) [reply kyo-core/jvm/src/main/scala/kyo/fibersPlatformSpecific.scala:7]: All good to go then 😸

**Comment by fwbrasil** [comment]
cool! Thanks for the PR :)

**Pull request merged**